MVP of Accessibility/UI automation support

grokys opened the pull request:
## What does the pull request do?

Adds a Minimum Viable Implementation of Accessibility/UI automation to Avalonia along with the start of some integration tests using Appium.

There is still a lot missing here, but the PR is already huge so I'd like to propose getting what's here merged and improving it in smaller PRs.

The API tries to stick as closely to the WPF/UWP API as possible, while adding a few improvements.

## Platforms

- [x] Windows (UIA)
- [x] MacOS (NSAccessibility)
- [ ] Linux (AT-SPI2)
  - Can't get Tmds.DBus to work with this
  
## Providers

- [ ] IDockProvider	
- [x] IExpandCollapseProvider	
- [ ] IGridItemProvider	
- [ ] IGridProvider	
- [x] IInvokeProvider	
- [ ] IItemContainerProvider	
- [ ] IMultipleViewProvider	
- [x] IRangeValueProvider	
- [x] IRootProvider (`IWindowProvider` in WPF)
- [x] IScrollItemProvider	
- [x] IScrollProvider	
- [x] ISelectionItemProvider	
- [x] ISelectionProvider	
- [ ] ISynchronizedInputProvider	
- [ ] ITableItemProvider	
- [ ] ITableProvider	
- [ ] ITextProvider/ITextRangeProvider	
  - I have an implementation of this, but the Windows model is so different from the MacOS model for text accessibility that it will need some more work
- [x] IToggleProvider	
- [ ] ITransformProvider	
- [x] IValueProvider	
- [ ] IVirtualizedItemProvider	

## Controls

- [ ] Calendar
- [ ] CaptionButton
- [ ] TitleBar
- [ ] DatePicker
- [ ] TimePicker
- [ ] MenuFlyout
- [ ] Notification
- [ ] NumericUpDown
- [ ] ScrollBar
- [ ] TabStrip
- [ ] ItemsRepeater
- [x] Button
- [ ] ButtonSpinner
- [ ] Carousel
- [x] CheckBox
- [x] ComboBox
- [x] ContextMenu
- [ ] Expander
- [x] Image
- [x] ItemsControl
- [ ] Label
- [x] ListBox
- [ ] MaskedTextBox
- [x] Menu
- [x] ProgressBar
- [x] RadioButton
- [x] RepeatButton
- [x] ScrollViewer
- [x] Slider
- [ ] Spinner
- [ ] SplitView
- [x] TabControl
- [x] TextBlock
- [x] ToggleSwitch
- [ ] TreeView
- [x] Window
- [ ] DataGrid

## Integration Tests

- Button
- CheckBox
- ComboBox
- ListBox
- Menu
- NativeMenu

Integration tests are currently using Appium in order to only need to write one set for Windows/MacOS. However Appium is very buggy on Windows/MacOS and also quite limited:

- WinAppDriver is [currently unmaintained](https://github.com/microsoft/WinAppDriver/issues/1550) and has problems doing basic stuff like Ctrl+Click
- appium-mac2-driver is maintained but in an even worse state - it too can't mange things like Ctrl+Click, or even __closing windows__ without crashing
- There doesn't seem to be any support for listening to a11y events in the C# client

Maybe that we need to write platform-specific integration tests in the long run, but for now lets see how far Appium can take us.

## Fixed issues

Fixes #585

grokys commented:
@jkoritzinsky in #585 it sounded like you had some ideas for this, and sounds like it's not really along the lines of what you were thinking (instead it's loosely based on WPF's automation system). Is this the right direction do you think?

MarchingCube commented:
To sum up my initial testing - I was able to use https://github.com/microsoft/WinAppDriver to interact with ControlCatalog running on top of this branch. I was able to navigate between tab pages and click buttons. I was mostly using names and type names to find things. I can see that having support for automation id will be useful: https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.automation.automationproperties.automationidproperty?view=winrt-19041

grokys commented:
@MarchingCube `AutomationId` should already be implemented: https://github.com/AvaloniaUI/Avalonia/pull/5177/files#diff-b209770a9453295e585e6281522b977f046b4fc2a9ac0bbdb9be3fe0dc0e2fbeR26-R32

MarchingCube commented:
@grokys Right, I didn't realize that it is an attached property. I will try testing again sometime soon.

grokys commented:
As I keep working on this the more I hate the WPF/UWP `AutomationPeer` API... You end up having to build three trees: `Logical Tree` -> `AutomationPeer tree` -> `OS Automation Node Tree` and keep them synchronized. For what end?

The more I look at the OSX NSAccessibility API the more I feel jealous. Compare the workflow in WPF and OSX to make a control an accessible button.

WPF:

- Create a `MyButtonAutomationPeer`
- Implement the `IInvokeProvider` interface on the peer
- Override `OnCreateAutomationPeer` on `MyButton` to return a `MyButtonAutomationPeer`

OSX:

- Implement [`NSAccessibilityButton`](https://developer.apple.com/documentation/appkit/nsaccessibilitybutton?language=objc) on `MyButton`

Does anyone have any experience of these APIs? Is there an advantage to the WPF/UWP way of doing it?

grokys committed:
Initial implementation of UI automation.

Follows WPF/UWP API as closely as possible. Limited to win32 right now. Broken in many places.

grokys committed:
Started adding integration tests.

grokys committed:
Started adding ComboBox integration tests.

And fixed some issues with ComboBox/popups.

grokys committed:
Make integration tests ordered.

grokys committed:
Add CheckBox tests.

grokys committed:
Use Name for AutomationId.

grokys committed:
Implemented LabeledBy.

grokys committed:
Added access/accelerator key support.

And start adding integration tests for menu items.

grokys committed:
Notify UIA that a window has been destroyed.

grokys committed:
Use Group as generic control type.

Seems to be what best fits, and what e.g. chrome uses.

grokys committed:
Add context menu automation peer.

And fixed NRE in MenuItemAutomationPeer.

grokys committed:
Don't use C#9 features.

Causing the build to fail on OSX/Linux

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Make root automation node come from OS.

This will be needed for OSX.

grokys committed:
Initial implementation of OSX automation.

grokys committed:
Use correct peer for TabItems.

grokys committed:
Handle children changing.

And map all the automation control types to roles.

grokys committed:
More OSX accessibility implementation.

grokys committed:
Implemented a11y focus on OSX.

grokys committed:
Fix a11y for popup windows.

- Don't show when creating a popup window with a11y enabled
- Raise focus changed notification on element that got focus instead of the window

grokys committed:
Use correct role for combobox items on OSX.

grokys committed:
More work on ComboBox OSX a11y.

Implement "Show Menu" and "Perform Press" actions.

grokys committed:
Expose max/min for slider controls etc.

chingucoding commented on src/Avalonia.Controls/Automation/AutomationProperties.cs:
```diff
@@ -0,0 +1,539 @@
+
+namespace Avalonia.Automation
+{
+    public static class AutomationProperties
```
A `HeadingLevel` property is missing, however it makes scanning documents a lot easier for people using assistive technology. Also, a `LandMarkType` property is missing but is very important for people with assistive technology.

grokys replied:
This was ported from WPF which doesn't have `HeadingLevel`/`LandmarkType` . I can add those.

chingucoding replied:
Those properties are present in UWP and the web (as semantic html). Honestly, I'm a bit surprised that WPF didn't have those.

chingucoding replied:
So should we add the `HeadingLevel` or `LandMarkType` properties? When using narrator on Windows, those are definitely helping people navigate UWP/WinUI apps more easily and those properties are also present in web accessibility.

grokys replied:
Yeah definitely, as I mention in the PR description though, this PR is already huge so stuff like this will be implemented later.

chingucoding commented on src/Avalonia.Controls/Automation/Peers/NoneAutomationPeer.cs:
```diff
@@ -0,0 +1,28 @@
+
+        protected override AutomationControlType GetAutomationControlTypeCore()
+        {
+            return AutomationControlType.Group;
```
This does not sound right, "None" content is a group? How about just removing it from UIA tree then instead of saying it is a group?

grokys replied:
The problem with completely removing them from the OS-level automation tree is that it breaks the `AutomationPeer` -> `AutomationNode`  1:1 mapping and means that we have to have _another_ level of abstraction to handle parent/child relationships in the case where peers don't map directly to OS automation nodes. We've already got 3 trees to keep in sync so at least if the nodes in the "peer" tree match 1:1 with the nodes in the OS tree it reduces some complexity.

Win32 and OSX provide a facility for "hiding" controls at the automation node level (`UIA_IsControlElementPropertyId`/[`accessibilityElement`](https://developer.apple.com/documentation/appkit/nsaccessibility/1535002-accessibilityelement?language=objc)), which looked to me to be sufficient?

Maybe instead of using `Group` we should have a new `AutomationControlType.None` type which will be mapped to the most relevant type for "ignored" elements at the OS level?

chingucoding replied:
Ah I see. In that case, group is fine, having a "none" control type would be really cool, however I'm not sure how much of a benefit there would be.

chingucoding commented on src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs:
```diff
@@ -17,6 +18,7 @@ public partial class WindowImpl
         protected virtual unsafe IntPtr AppWndProc(IntPtr hWnd, uint msg, IntPtr wParam, IntPtr lParam)
         {
             const double wheelDelta = 120.0;
+            const long UiaRootObjectId = -25;
```
Is the -25 arbitrary here?

grokys replied:
This is from the Win32 API: https://github.com/tpn/winsdk-10/blob/master/Include/10.0.16299.0/um/UIAutomationCoreApi.h#L52

chingucoding replied:
I see, good to know. Could add a comment for this, but I think it's fine as is.

chingucoding commented on tests/Avalonia.Controls.UnitTests/Automation/ControlAutomationPeerTests.cs:
```diff
@@ -0,0 +1,253 @@
+
+namespace Avalonia.Controls.UnitTests.Automation
+{
+    public class ControlAutomationPeerTests
```
Should we test that properties trickle down, e.g. AutomationProperties.Name ?

grokys replied:
I've not implemented `Name` yet, how should it "trickle down"?

chingucoding replied:
With "trickle down" I was thinking about trying to use control properties to get the name. E.g. Button should try to use it's content as string to see if it get's that name then. I think the following order is good:

1. Try `AutomationProperties.Name` to see if there is a name
2. Try to defer name from content (Button content or text of a TextBlock)
3. Try super.GetNameCore()
4. Return empty string as there was no way of getting actual name

grokys replied:
Actually, seems like I have implemented this: https://github.com/AvaloniaUI/Avalonia/blob/feature/ui-automation/src/Avalonia.Controls/Automation/Peers/ControlAutomationPeer.cs#L91

I think this should work ask you expect, where derived `AutomationPeer`s try to get the name from e.g. the button content, and if they fail they call this base implementation which checks the property.

chingucoding replied:
I see, yes that should work as expected. Seems fine to me.

chingucoding commented on src/Avalonia.Controls/Automation/AutomationProperties.cs:
```diff
@@ -0,0 +1,539 @@
+        /// <summary>
+        /// Defines the AutomationProperties.Name attached attached property.
+        /// </summary>
+        public static readonly AttachedProperty<string> NameProperty =
```
It doesn't seem any peer respects this property. If not even the base peer or control peer respects this, why include this property?

grokys replied:
I still need to implement quite a few properties from this class.

grokys replied:
See https://github.com/AvaloniaUI/Avalonia/blob/feature/ui-automation/src/Avalonia.Controls/Automation/Peers/ControlAutomationPeer.cs#L91

marcelwgn reviewed (commented):
I think the UIA API from WPF/UWP are good and I think the API you have so far (besides my remarks/comments) is good. One thing you could change is that the `AutomationControlType` could also be set from XAML. Some peers only seem to exist to override this peer, this might not be ideal. In contrast, the Web A11y API (aria) only works using HTML properties, so it might be worth exploring moving things that don't need to be done from code behind into XAML (e.g. `AutomationControlType`).

grokys commented:
Thanks so much for the review @chingucoding - as I've mentioned I'm new to this stuff so getting some input from someone with experience in this area is invaluable.

> One thing you could change is that the AutomationControlType could also be set from XAML. Some peers only seem to exist to override this peer, this might not be ideal.

Sounds like a good idea - so something like `<Button AutomationProperties.AutomationControlType="foo">` which when set overrides the control type in the peer?

> In contrast, the Web A11y API (aria) only works using HTML properties, so it might be worth exploring moving things that don't need to be done from code behind into XAML (e.g. AutomationControlType).

Yep, peers definitely do seem heavyweight from my point of view compared to HTML area properties! Any other ideas for things we could do directly from XAML?

grokys commented on src/Avalonia.Controls/Automation/Peers/AutomationPeer.cs:
```diff
@@ -0,0 +1,259 @@
+
+namespace Avalonia.Automation.Peers
+{
+    public enum AutomationControlType
```
Yeah this list comes from the WPF API. Should we just replace this list with values from the ARIA roles if that's the common subset? Should we keep the WPF/win32 naming or use the ARIA naming? Perhaps that latter would make more sense from an x-plat point of view.

chingucoding replied:
I think sticking with the WPF/Win23/UWP naming is easier since folks familiar with XAML are more familiar with those names. Developers coming from the Web will have to learn XAML first anyway, having some web naming in here might seem weird then.

grokys commented on src/Avalonia.Controls/Automation/Peers/AutomationPeer.cs:
```diff
@@ -0,0 +1,259 @@
+            Node.PropertyChanged(automationProperty, oldValue, newValue);
+        }
+
+        protected virtual string GetLocalizedControlTypeCore()
```
Just not done yet. Would it be best to use satellite assembles for localizing the names, or just hardcode the names for all languages in code/resources in the main assembly?

chingucoding replied:
Probably hardcoding is fine? Not sure how localization works with Avalonia so whatever is the best way of doing it in Avalonia might be best here. But as a developer I would expect this assembly to be able to return translated control types.

grokys commented:
In addition, I have an initial implementation for OSX that I need to merge into this PR. Are you familiar with a11y on OSX at all?

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Fix AccessText after merge.

grokys commented:
Added the initial OSX implementation.

chingucoding commented:
> Thanks so much for the review @chingucoding - as I've mentioned I'm new to this stuff so getting some input from someone with experience in this area is invaluable.
> 

Happy that I can help!

> > One thing you could change is that the AutomationControlType could also be set from XAML. Some peers only seem to exist to override this peer, this might not be ideal.
> 
> Sounds like a good idea - so something like `<Button AutomationProperties.AutomationControlType="foo">` which when set overrides the control type in the peer?

Yes, that is what I was thinking might be a good addition.

> > In contrast, the Web A11y API (aria) only works using HTML properties, so it might be worth exploring moving things that don't need to be done from code behind into XAML (e.g. AutomationControlType).
> 
> Yep, peers definitely do seem heavyweight from my point of view compared to HTML aria properties! Any other ideas for things we could do directly from XAML?

I think almost all other things are already a property. One thing that could be helpful is being able to set "IsControlElement" and "IsContentElement" from XAML. In UWP/WinUI you have the [AccessibilityView](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.automation.peers.accessibilityview?view=winrt-20348) property that you can use for this. Maybe something similar can be considered?


> In addition, I have an initial implementation for OSX that I need to merge into this PR. Are you familiar with a11y on OSX at all?

Unfortunately not, sorry.

sid-dahiya commented:
@grokys , I was able to test out Voice Over on the test ControlCatalog.NetCore app in macOS Big Sur (`11.4 (20F71)`). It looks like the voice overs are working correctly. I'm now trying to perform a sanity check by integrating a nuget with your changes into our app as well.

So far, I'm following the conversation above in the comments. Please let me know if there are other scenarios you'd like validated as well.

jonthysell commented:
First off, full disclosure, I work for MSFT, I've worked on UWP XAML, so I probably can't
contribute code. But I've worked on accessibility, so I love that this PR exists, and I figured I'd put in my two cents.

In React Native Windows (my current work), we also ran into the problem of needing to create custom automation peers to map to arbitrary JS-based React Native controls. Rather than have to create a whole hierarchy of peers, I created a single DynamicAutomationPeer uses attached properties set on the peer XAML control to decide which automation providers it reports to UIA. It solved a lot of problems with having to override OnCreateAutomationPeer. It was even proposed back into XAML proper: https://github.com/microsoft/microsoft-ui-xaml/issues/1349

grokys commented:
@sid-dahiya thank you! I think the main doubt I had on OSX was that navigation using VoiceOver didn't seem to work correctly, though I'm not sure why. I couldn't find the a11y API which controlled that...

@jonthysell at first glance I _much_ prefer that API! If you were implementing a11y for Avalonia yourself would you keep the current peer API around or would you move to that API?

jonthysell commented:
@grokys Obviously there's a certain amount of infrastructure you'll need - an enum of types, provider interfaces, the automation peer tree, and probably some "base" (abstract?) automation peer that can have a UIA type, name, id, and can be queried for its providers.

Then you can really have two options - the existing paradigm that WPF/UWP follows, where almost every control class has a matched automationpeer class which overrides the specifics through class inheritance. This has the benefits of being a known working system, that also lets you create UIA trees which diverge from the visual tree. However it also has the downside that the *only* way to customize the a11y behavior of a control is to subclass the control just to override `OnCreateAutomationPeer`, and you also probably have to create a whole new peer class as well and maybe not be able to share implementation code nicely.

With the `DynamicAutomationPeer` approach, you can create something that's more akin to ARIA, where any control's a11y behavior can be customized just by setting (attached) properties. In React Native, we use this because developers create custom user controls from scratch in JavaScript, so we give them a native panel (View, equivalent to an HTML div) that has this dynamic peer. Now they can define a11y behavior (setting type, binding to a11y actions like Invoke) by setting properties on that panel which contains their control.

You'll probably always want `OnCreateAutomationPeer` available for the cases where it's needed, and for developers with complicated control requirements who may already be used to that solving a11y like they did in WPF/UWP.

However, you may find that, in implementing the automation peers for the set of controls "built-in" to avalonia, that it's easier to have one very flexible, shared DynamicAutomationPeer class driven by properties. You could have the base Control class's `OnCreateAutomationPeer` return a DynamicAutomationPeer by default, and rely on setting properties within each control's code / template. Then developers have the option of overriding/extending the built-in Control's behavior by setting properties rather than subclassing, more easily defining the a11y behaviors for their own templated controls (by setting properties directly in the template), not to mention getting around any potential issues with being unable to subclass sealed controls in the ecosystem (maybe not a problem with Avalonia, but sometimes a problem in WPF/UWP XAML).

Finally, you could even a *different* namespace of a11y concepts/terms in your properties, which is what we do in React Native Windows. React Native already had its own set of a11y properties on controls based on a least common denominator of ios/Android properties, and we could do the mapping to UIA terms within the DynamicAutomationPeer. So, for example, you could expose literal ARIA property names on avalonia controls, then have DynamicAutomationPeer translate that into UIA for Windows, and whatever the equivalents on OSX, etc. Or create your own platform-agnostic Avalonia a11y properties that can be mapped under the hood. I'm not sure where (if anywhere) you're planning on abstracting away platform specifics from control developers, of if you expect control developers to have to define a11y behavior for every target platform themselves.

Anyway, strictly speaking, if you want to leverage and align as much as possible with WPF, it's a long-standing system "proven in the field" as it were. But diverging slightly and adding a DynamicAutomationPeer (especially early on) may prove useful. I certainly enjoy some of the places where avalonia has diverged from WPF (like in styles).

sid-dahiya commented:
@grokys DataGrid seems to be the only control missing automation peer support. Could that also be added?

sid-dahiya commented:
> @grokys DataGrid seems to be the only control missing automation peer support. Could that also be added?

I tried to add `DataGrid` peer locally, but can't get it to work. Originally, I tried to do it on the `DataGrid` control, but soon realized that for the best experience, it should be on the `DataGridCell` control since that is where a user would be when tabbing through. Now, since the `DataGridCell` seems to support any control (?), I decided to try and identify it's Content and call the peers on that. That didn't seem to work either (or most probably, I was not calling the content correctly).

@grokys thoughts on how to add DataGrid support? I noticed it is in the `AutomationControlType` list.

grokys commented:
@sid-dahiya thanks for looking into it! As I mentioned on the other thread, I'm not going to be able to work on this for a while so help is appreciated.

I don't know much about `DataGrid` (it was ported from Silverlight toolkit before WPF was open-sourced) but does the WPF source help here? https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Automation/Peers/DataGridAutomationPeer.cs

sid-dahiya commented:
> @sid-dahiya thanks for looking into it! As I mentioned on the other thread, I'm not going to be able to work on this for a while so help is appreciated.
> 
> I don't know much about `DataGrid` (it was ported from Silverlight toolkit before WPF was open-sourced) but does the WPF source help here? https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Automation/Peers/DataGridAutomationPeer.cs
> 
> Note that even though it's on `AutomationControlType` the platform peers for data grid types haven't been implemented yet.

Sounds good, let me take a poke at it.

ShrutiJaiswal1494 commented:
Hi,

I am using a locally generated NuGet package for my application with all the accessibility changes in this [branch](https://github.com/AvaloniaUI/Avalonia/tree/feature/ui-automation). On Windows (Windows 10 and 11), when I click on the combo box control or the windows maximize button with the narrator or other screen readers on, the app crashes with the below errors. Also, there is an interesting thing happening with the combo box that it will work fine 1/10 times.
Any help on how to resolve these errors would really be appreciated. Also, it would be great if @grokys can look into this and help me get a solution. Thanks in advance.

**Error for ComboBox**

![image](https://user-images.githubusercontent.com/79670853/134975616-6e1867fc-c12f-43dc-be2d-e3d987ee6d7b.png)

Full Stack trace:
```
   at Avalonia.Controls.WindowBase.OnCreateAutomationPeer(IAutomationNodeFactory factory) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/WindowBase.cs:line 277
   at Avalonia.Controls.Control.GetOrCreateAutomationPeer(IAutomationNodeFactory factory) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Control.cs:line 229
   at Avalonia.Automation.Peers.ControlAutomationPeer.GetOrCreatePeer(Control element) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Automation/Peers/ControlAutomationPeer.cs:line 46
   at Avalonia.Automation.Peers.PopupAutomationPeer.GetPopupRoot() in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Automation/Peers/PopupAutomationPeer.cs:line 49
   at Avalonia.Automation.Peers.PopupAutomationPeer.PopupOpenedClosed(Object sender, EventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Automation/Peers/PopupAutomationPeer.cs:line 42
   at Avalonia.Controls.Primitives.Popup.Open() in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Primitives/Popup.cs:line 483
   at Avalonia.Controls.Primitives.Popup.IsOpenChanged(AvaloniaPropertyChangedEventArgs`1 e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Primitives/Popup.cs:line 556
   at Avalonia.Controls.Primitives.Popup.<>c.<.cctor>b__22_1(Popup x, AvaloniaPropertyChangedEventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Primitives/Popup.cs:line 146
   at System.Reactive.Subjects.Subject`1.OnNext(T value) in /_/Rx.NET/Source/src/System.Reactive/Subjects/Subject.cs:line 147
   at Avalonia.AvaloniaObject.RaisePropertyChanged[T](AvaloniaPropertyChangedEventArgs`1 change) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 763
   at Avalonia.AvaloniaObject.SetAndRaise[T](AvaloniaProperty`1 property, T& field, T value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 705
   at Avalonia.Controls.Primitives.Popup.set_IsOpen(Boolean value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Primitives/Popup.cs:line 208
   at Avalonia.Controls.Primitives.Popup.<>c.<.cctor>b__22_3(Popup o, Boolean v) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/Primitives/Popup.cs:line 40
   at Avalonia.DirectProperty`2.InvokeSetter(IAvaloniaObject instance, BindingValue`1 value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/DirectProperty.cs:line 174
   at Avalonia.AvaloniaObject.SetDirectValueUnchecked[T](DirectPropertyBase`1 property, BindingValue`1 value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 844
   at Avalonia.Reactive.TypedBindingAdapter`1.OnNext(BindingValue`1 value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/Reactive/TypedBindingAdapter.cs:line 44
   at Avalonia.Reactive.BindingValueAdapter`1.OnNext(T value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/Reactive/BindingValueAdapter.cs:line 18
   at Avalonia.AvaloniaObject.RaisePropertyChanged[T](AvaloniaPropertyChangedEventArgs`1 change) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 765
   at Avalonia.AvaloniaObject.SetAndRaise[T](AvaloniaProperty`1 property, T& field, T value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/AvaloniaObject.cs:line 705
   at Avalonia.Controls.ComboBox.set_IsDropDownOpen(Boolean value) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/ComboBox.cs:line 112
   at Avalonia.Controls.ComboBox.OnPointerReleased(PointerReleasedEventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/ComboBox.cs:line 320
   at System.Reactive.Subjects.Subject`1.OnNext(T value) in /_/Rx.NET/Source/src/System.Reactive/Subjects/Subject.cs:line 149
   at Avalonia.Interactivity.EventRoute.RaiseEventImpl(RoutedEventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Interactivity/EventRoute.cs:line 148
   at Avalonia.Interactivity.EventRoute.RaiseEvent(IInteractive source, RoutedEventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Interactivity/EventRoute.cs:line 102
   at Avalonia.Interactivity.Interactive.RaiseEvent(RoutedEventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Interactivity/Interactive.cs:line 122
   at Avalonia.Input.MouseDevice.MouseUp(IMouseDevice device, UInt64 timestamp, IInputRoot root, Point p, PointerPointProperties props, KeyModifiers inputModifiers) in /Users/Desktop/repos/Avalonia/src/Avalonia.Input/MouseDevice.cs:line 307
   at Avalonia.Input.MouseDevice.ProcessRawEvent(RawPointerEventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Input/MouseDevice.cs:line 149
   at Avalonia.Input.InputManager.ProcessInput(RawInputEventArgs e) in /Users/Desktop/repos/Avalonia/src/Avalonia.Input/InputManager.cs:line 36
   at Avalonia.Win32.WindowImpl.AppWndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam) in C:\Users\source\repos\NugetAvalonia\Avalonia\src\Windows\Avalonia.Win32\WindowImpl.AppWndProc.cs:line 489
   at Avalonia.Win32.WindowImpl.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam) in C:\Users\source\repos\NugetAvalonia\Avalonia\src\Windows\Avalonia.Win32\WindowImpl.WndProc.cs:line 33
   at Avalonia.Win32.Interop.UnmanagedMethods.DispatchMessage(MSG& lpmsg)
   at Avalonia.Win32.Win32Platform.RunLoop(CancellationToken cancellationToken) in C:\Users\source\repos\NugetAvalonia\Avalonia\src\Windows\Avalonia.Win32\Win32Platform.cs:line 157
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken) in /Users/Desktop/repos/Avalonia/src/Avalonia.Base/Threading/Dispatcher.cs:line 62
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 117
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode) in /Users/Desktop/repos/Avalonia/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 148
   at Office.Telemetry.UI.Program.Main(String[] args) in C:\Users\source\repos\MyApps\DemoApp\DemoAppUI\Program.cs:line 11
``` 


**Error for Window Maximize Button**

![image](https://user-images.githubusercontent.com/79670853/134976383-387c0a13-66a1-4a4c-b7f1-9ca8bcd9d821.png)

Full Stack trace
```
   at System.StubHelpers.ObjectMarshaler.ConvertToNative(Object objSrc, IntPtr pDstVariant)
   at Avalonia.Win32.Interop.Automation.UiaCoreProviderApi.UiaRaiseAutomationPropertyChangedEvent(IRawElementProviderSimple provider, Int32 id, Object oldValue, Object newValue)
   at Avalonia.Win32.Automation.AutomationNode.PropertyChanged(AutomationProperty property, Object oldValue, Object newValue) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Windows\Avalonia.Win32\Automation\AutomationNode.cs:line 99
   at Avalonia.Automation.Peers.AutomationPeer.RaisePropertyChangedEvent(AutomationProperty automationProperty, Object oldValue, Object newValue) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Controls\Automation\Peers\AutomationPeer.cs:line 202
   at Avalonia.Automation.Peers.ControlAutomationPeer.OwnerPropertyChanged(Object sender, AvaloniaPropertyChangedEventArgs e) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Controls\Automation\Peers\ControlAutomationPeer.cs:line 192
   at Avalonia.AvaloniaObject.RaisePropertyChanged[T](AvaloniaPropertyChangedEventArgs`1 change) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Base\AvaloniaObject.cs:line 763
   at Avalonia.AvaloniaObject.RaisePropertyChanged[T](AvaloniaProperty`1 property, Optional`1 oldValue, BindingValue`1 newValue, BindingPriority priority) in C:\Users\shjaiswal\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Base\AvaloniaObject.cs:line 674
   at Avalonia.AvaloniaObject.SetAndRaise[T](AvaloniaProperty`1 property, T& field, T value) in C:\Users\shjaiswal\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Base\AvaloniaObject.cs:line 704
   at Avalonia.Visual.Avalonia.VisualTree.IVisual.set_TransformedBounds(Nullable`1 value) in C:\Users\shjaiswal\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Visuals\Visual.cs:line 291
   at Avalonia.Rendering.SceneGraph.SceneBuilder.Update(DrawingContext context, Scene scene, VisualNode node, Rect clip, Boolean forceRecurse) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Visuals\Rendering\SceneGraph\SceneBuilder.cs:line 246
   at Avalonia.Rendering.SceneGraph.SceneBuilder.Update(Scene scene, IVisual visual) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Visuals\Rendering\SceneGraph\SceneBuilder.cs:line 91
   at Avalonia.Rendering.DeferredRenderer.UpdateScene() in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Visuals\Rendering\DeferredRenderer.cs:line 641
   at Avalonia.Rendering.DeferredRenderer.Paint(Rect rect) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Visuals\Rendering\DeferredRenderer.cs:line 222
   at Avalonia.Controls.TopLevel.HandlePaint(Rect rect) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Avalonia.Controls\TopLevel.cs:line 334
   at Avalonia.Win32.WindowImpl.AppWndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Windows\Avalonia.Win32\WindowImpl.AppWndProc.cs:line 363
   at Avalonia.Win32.WindowImpl.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam) in C:\Users\source\repos\AvaloniaNuget\Avalonia\src\Windows\Avalonia.Win32\WindowImpl.WndProc.cs:line 30

```

grokys commented:
Hi @ShrutiJaiswal1494 - I PM'd you on gitter a few months back to chat about this, but I don't think you've seen it. Might be easier if we took this to chat?

ShrutiJaiswal1494 commented:
@grokys sorry, I missed your message. Replied to you there.

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Update ApiCompatBaseline.txt

grokys committed:
Try to fix problem with ComboBox.

Only the OS can create the root automation peer, so we need to take care to not accidentally try to create it from `PopupAutomationPeer`.

grokys committed:
Revert "Try to fix problem with ComboBox."

This reverts commit c34a825bf5507a2ea87c9afd9e6cea3ff53c7594.

grokys committed:
Decouple automation peers from platform nodes.

Now peers are entirely unaware of the platform implementation.

grokys committed:
Create root automation nodes like others.

grokys committed:
Use non-obsolete interface.

And remove debug code.

grokys committed:
Added AutomationControlType.None.

grokys committed:
Reinstate ControlAutomationPeerTests.

BaseHead commented:
So is the Windows Narrator supposed to be working on Win10?  or not yet?  cuz I only here it when our application is selected but never when I mouse over any object/control inside the application.

grokys commented:
@BaseHead not yet unfortunately. This PR is still a work in progress and hasn't been merged to master yet.

grokys committed:
Allow peers to dynamically expose implemented interfaces.

grokys committed:
Added AccessibilityView and ControlType.

- `AutomationProperties.AccessibilityView`: from UWP - control's element's visibility in automation tree
- `ControlTypeOverride`: from UWP proposal - overrides the automation control type specified in peer

grokys committed:
Use new properties to remove some peers.

Simple peers which only change the accessibility view or control type can now use the new `AutomationProperties`.

grokys committed:
Query correct interface.

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Move ChildrenChanged to AutomationPeer.

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Make OSX a11y work again.

Pretty major refactor of that code.

grokys committed:
Try to prevent leaks.

grokys committed:
Stop warning.

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Update IntegrationTestApp to net6.0.

grokys committed:
Trying to use a shared project for win32/mac integration tests.

grokys committed:
WIP: Bundle IntegrationTestApp.

grokys committed:
Added script to create bundle.

grokys committed:
Start making automation tests work on macOS.

grokys committed:
Correctly raise layout changed notification (?)

grokys committed:
Get CheckBox automation tests passing on macOS.

grokys committed:
Rename integration tests now they're not win32-only.

grokys committed:
Skip accelerator key test on OSX.

OSX doesn't support accelerator keys.

grokys committed:
Raise focus changed on new window if it has focus.

grokys committed:
Make integration tests pass on MacOS.

grokys committed:
Use ListItemAutomationPeer for TabItem.

grokys committed:
Implement isAccessibilitySelected.

grokys committed:
Added a NativeMenu test for macOS.

grokys committed:
Refactored Menu integration tests.

grokys committed:
Remove rogue character.

grokys committed:
Remove faulty cast.

grokys committed:
Refactored ComboBox integration tests.

grokys committed:
Trying to get stuff working on macOS.

Man, appium for mac sucks.

grokys committed:
Skip keyboard interaction tests on macOS.

They send appium for mac crazy.

grokys committed:
Added a few ListBox tests.

grokys committed:
Sigh. Appium sucks.

grokys commented:
I think this is now nearing the stage of MVP. I added some attached properties for setting things like `ControlType` and `AccessibilityView` as suggested upthread, though support for these has been added to `AutomationPeer` instead of a separate `DynamicAutomationPeer`.

I have also added the start of some integration tests using Appium. Unfortunately the Windows and Mac drivers for Appium are very buggy _indeed_, so I'm not sure how feasible it is to use them for our integration tests in the long-term. 

@ShrutiJaiswal1494 if you're still interested, I pinged you on gitter a few weeks ago but pinging you here in case you didn't see.

grokys committed:
Added integration tests readme.

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Add additional macOS prerequisite.

grokys commented:
Ok, this is ready for review as an MVP.

grokys committed:
Updated ApiCompat.

pr8x commented on native/Avalonia.Native/src/OSX/AvnString.mm:
```diff
@@ -153,3 +153,20 @@ virtual HRESULT Get(unsigned int index, IAvnString**ppv) override
+            return [NSString stringWithUTF8String:p];
+        }
+        
+        s->Release();
```
Wait... This never releases when the `NSString` returns above?

grokys replied:
Yeah, looks like a mistake 👍

chingucoding commented on src/Avalonia.Controls/Automation/Peers/PopupRootAutomationPeer.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    public class PopupRootAutomationPeer : WindowBaseAutomationPeer
+    {
+        public PopupRootAutomationPeer(PopupRoot owner)
```
What is this AutomationPeer used for? Only for tracking the focus around popups?

grokys replied:
Yeah pretty much. It's almost the same as `WindowAutomationPeer` except that popups don't have a title and aren't control/content elements.

chingucoding commented on src/Avalonia.Controls/Automation/Peers/TextBoxAutomationPeer.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public class TextBoxAutomationPeer : ControlAutomationPeer, IValueProvider
+    {
+        public TextBoxAutomationPeer(TextBox owner)
```
Should this peer read out the placeholder/watermark textif the TextBox has no text entered?

grokys replied:
Yep, probably! It also needs to implement `ITextProvider` so a fair bit of work will need to be done on this class in a future PR.

jmacato committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Throw when interface not supported.

grokys committed:
Implement provider members as explicit interface members.

Previous change showed up a problem where incorrect interface was getting called for `IsReadOnly`.

marcelwgn reviewed (approved):
I think this is heading in the right direction. There are still thinks that need to be implement but that is expected since this is the MVP. In my opinion, this would be good API wise (didn't check all implementations exactly).

grokys committed:
Fix leak in GetNSStringAndRelease.

grokys committed:
Don't run integration tests under ncrunch.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0017726-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Changed by VS.

grokys committed:
Remove non-existent project.

grokys committed:
Cast to interface instead of concrete type.

`PlatformImpl` may be a `ValidatingTopLevelImpl`.

grokys committed:
Use correct interface for root window impl.

grokys committed:
Merge branch 'master' into feature/ui-automation

grokys committed:
Fix nullable errors after merge.

And remove now-unneeded `#nullable enable` statements.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018922-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018922-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into feature/ui-automation

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018944-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018944-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys committed:
Merge branch 'master' into feature/ui-automation

danwalmsley reviewed (approved):
Lgtm yolo

webczat commented:
is usage of ATK instead of AT-SPI a viable option for adding linux accessibility? might be easier... or not

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018988-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018988-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

zijwij-byxmyq-1wIxre reviewed (commented):
ZDZ
